lib/template/tests/test_template_compile.c

```diff
@@ -442,49 +360,44 @@ test_template_function_bad2(void)
                            default_value = NULL, macro = M_NONE, type = LTE_MACRO, msg_ref = 0);
 }
 
-static void
-test_template_function_bad3(void)
+Test(template_compile, test_template_function_bad3)
 {
   assert_failed_template_compile("$(hello \"This is an unclosed quoted string)",
                                  "Invalid template function reference, missing function name or imbalanced '(', error_pos='8'");
   assert_compiled_template(text = "error in template: $(hello \"This is an unclosed quoted string)", default_value = NULL,
                            macro = M_NONE, type = LTE_MACRO, msg_ref = 0);
 }
 
-static void
-test_unknown_function(void)
+Test(template_compile, test_unknown_function)
 {
   assert_failed_template_compile("$(unknown function)", "Unknown template function \"unknown\"");
   assert_compiled_template(text = "error in template: $(unknown function)", default_value = NULL, macro = M_NONE,
                            type = LTE_MACRO, msg_ref = 0);
 }
 
 static void
-test_template_compile_negativ_tests(void)
+setup(void)
 {
-  TEMPLATE_TESTCASE(test_invalid_macro);
-  TEMPLATE_TESTCASE(test_invalid_subst);
-  TEMPLATE_TESTCASE(test_template_function_bad1);
-  TEMPLATE_TESTCASE(test_template_function_bad2);
-  TEMPLATE_TESTCASE(test_template_function_bad3);
-  TEMPLATE_TESTCASE(test_unknown_function);
-}
-
-int main(int argc, char **argv)
-{
-  msg_init(FALSE);
-
   configuration = cfg_new_snippet();
   log_msg_registry_init();
   log_template_global_init();
   plugin_register(&configuration->plugin_context, &hello_plugin, 1);
 
-  test_template_compile_macro();
-  test_template_compile_value();
-  test_template_compile_func();
-  test_template_compile_negativ_tests();
+  template = log_template_new(configuration, NULL);
+  start_grabbing_messages();
+  app_startup();
+}
 
+static void
+teardown(void)
```

Comment:
There is a known bug in Criterion: https://github.com/Snaipe/Criterion/issues/242 ; if the teardown crashes it won't report it as an error (at least older criterion versions), your test actually crashes in teardown, you can check it by running the testcase with extra parameters: `./lib/template/tests/test_template_compile --verbose -j1` and the following warning is displayed for me: `Warning! The test `template_compile::test_value_without_braces` exited during its setup or teardown.`

Reply:
Hi @Kokan ,

Thanks for pointing it out. I am looking into it ...